Differentiate views on the frontend (public dataset navigability)

pavish opened the pull request:
Related to https://github.com/mathesar-foundation/mathesar/issues/4824
Related to https://github.com/mathesar-foundation/mathesar/issues/4926

**Details**
- Returns information on whether a relation is a view from the backend.
- Differentiates views on the frontend in the schema page.
- Updates i18n text in areas that are immediately visible.
- Irrelevant sections in the Table tab in the inspector are removed for views.

**Screenshots**
- Views are differentiated with a new icon and an additional tag in the schema page. I added the tag since the icon didn't seem sufficient enough to help differentiate views from tables.
  - <img width="894" height="798" alt="Screenshot 2025-11-20 at 6 43 15 PM" src="https://github.com/user-attachments/assets/14f8afad-6eb0-4fea-b14f-f6e3668c695d" />
  - <img width="287" height="270" alt="Screenshot 2025-11-20 at 6 44 33 PM" src="https://github.com/user-attachments/assets/ce44fffb-c637-4305-83fb-4a451516b096" />
- Text is updated in locations that are visible the most. Schema page, Table inspector etc., Some sections that are not relevant to views are removed in the table inspector (only in the Table tab).
  - <img width="648" height="179" alt="Screenshot 2025-11-20 at 6 45 06 PM" src="https://github.com/user-attachments/assets/db6477b6-85b7-4814-8dae-4dee91b2107e" />
  - <img width="1251" height="775" alt="Screenshot 2025-11-20 at 6 45 31 PM" src="https://github.com/user-attachments/assets/a68d76e8-cb18-44c5-8ee9-8dec70214c07" />

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

pavish committed:
Update table section titles to Tables & Views

pavish committed:
Add relkind information to the get_table_info function

pavish committed:
Represent view with a different icon

pavish committed:
Update text in table card for views

pavish committed:
Merge branch 'develop' of https://github.com/mathesar-foundation/mathesar into views_fe

pavish committed:
Add new custom icon for view

pavish committed:
Add tags to table card to indicate views

pavish committed:
Update i18n for views in all modals in table card

pavish committed:
Update i18n text for views in table tab in inspector

pavish committed:
Update batchRun in table update requests so that errors are properly thrown

pavish commented:
@seancolsen Please review frontend code & UX/visual changes.
  - https://github.com/mathesar-foundation/mathesar/pull/4992/commits/9945fd3caadae624dc0b6713df84ba6df5e07fdc requires more attention since updating a table now throws an error during failure instead of silently failing. I updated this since modifying comments for views was silently failing. I checked for possible regressions, I think we're good, but it would help to get another set of eyes on this.

@zackkrida Please review UX/visual changes.
@mathemancer Please review backend code.

Note that this PR aims to do the minimal amount of work needed for the public dataset navigability project & basic support for views on the frontend.

mathemancer reviewed (approved):
The backend code looks great!

zackkrida reviewed (approved):
@pavish this looks good to merge to me. I have some small preferences about the icon, color, and "view" tag in the table cards but those are subjective personal opinions. I'll share them anyway:

- I'd prefer to see a more visually distinct icon (like a plain eye). The current icon is too similar.
- In the same way, I'd prefer a distinct color. I understand the intent behind using a similar color but I think it would actually be more helpful to have a stronger identity for views.
- Both of these changes would obviate the need for the "View" tag in the table cards

Quick example and  diff:

<img width="1105" height="551" alt="Image" src="https://github.com/user-attachments/assets/67d08d1e-e839-4c36-bfb0-10189afff54b" />

```diff
diff --git a/mathesar_ui/src/App.svelte b/mathesar_ui/src/App.svelte
index ac1e28843..658d5f235 100644
--- a/mathesar_ui/src/App.svelte
+++ b/mathesar_ui/src/App.svelte
@@ -49,7 +49,7 @@
     'schema': $salmon,
     'database': $amethyst,
     'table': $pumpkin,
-    'view': hsl(hue($pumpkin), 70%, 45%),
+    'view': hsl(220, 65%, 55%),
     'column': hsl(hue($salmon), 40%, 60%),
     'record': hsl(296, 35%, 45%),
     'record-fk': hsl(296, 70%, 45%),
diff --git a/mathesar_ui/src/icons/index.ts b/mathesar_ui/src/icons/index.ts
index 59997ba4a..5a15321bc 100644
--- a/mathesar_ui/src/icons/index.ts
+++ b/mathesar_ui/src/icons/index.ts
@@ -30,6 +30,7 @@ import {
   faExpand,
   faExternalLink,
   faExternalLinkAlt,
+  faEye,
   faFile,
   faFileAlt,
   faFileArchive,
@@ -101,7 +102,6 @@ import {
   faWandSparkles,
   faXmark,
 } from '@fortawesome/free-solid-svg-icons';
-
 import type { IconProps } from '@mathesar-component-library/types';
 
 import {
@@ -118,7 +118,6 @@ import {
   permissionsIcon,
   tableIcon,
   treeChildNodeArrowIcon,
-  viewIcon,
 } from './customIcons';
 
 /**
@@ -233,7 +232,7 @@ export const iconSorting: IconProps = { data: faSort };
 export const iconTable: IconProps = { data: tableIcon };
 export const iconInspector: IconProps = { data: faInfo };
 export const iconTableLink: IconProps = { data: faLink };
-export const iconView: IconProps = { data: viewIcon };
+export const iconView: IconProps = { data: faEye };
 export const iconTechnicalExplanation: IconProps = { data: faHammer };
 export const iconTreeChildNodeArrow: IconProps = {
   data: treeChildNodeArrowIcon,
diff --git a/mathesar_ui/src/pages/schema/TableCard.svelte b/mathesar_ui/src/pages/schema/TableCard.svelte
index 45593a662..69fcb7b60 100644
--- a/mathesar_ui/src/pages/schema/TableCard.svelte
+++ b/mathesar_ui/src/pages/schema/TableCard.svelte
@@ -90,7 +90,7 @@
   description={table.description ?? undefined}
   icon={tableIcon}
   {pendingMessage}
-  tags={isView ? [$_('view')] : undefined}
+  tags={undefined}
   primary
   cssVariables={{
     '--EntityListItem__accent-color': accentColor,
```

kgodey commented:
Can we show views in a separate list above tables (same styling is fine), but only show that section if there are any views? This changes the header to "Tables & Views" regardless of whether the user has views, which is extremely confusing, especially since views aren't mentioned anywhere in Mathesar and there's no way to create them in Mathesar/

kgodey reviewed (changes requested):
Please see comment in PR.

seancolsen reviewed (approved):
Code looks good.

I noticed the following functionality that doesn't work but is present (and usable) in the UI:

- Rename view
- Delete view
- Delete column
- Add column
- Alter column data type
- Extract columns
- Set column default value

The user gets weird errors when trying any of these things on views. Maybe that's okay. Ideally we'd disable all this UI. I'm not sure how important that is right now though.

I don't have any opinions on the colors or icons.

pavish committed:
Merge branch 'develop' into views_fe

pavish committed:
Change header dynamically, sort views at the end and make view colors more distinct

pavish commented:
I've addressed some concerns raised by @kgodey and @zackkrida in https://github.com/mathesar-foundation/mathesar/pull/4992/commits/e672df9d9e94fe5d37fd6a9ff80cc141c392426f.
 - Views are placed at the end of the list.
    - Since views and tables take up the same namespace in the db, and a table with the same name as a view cannot be created, I think it might be best to keep them in the same list.
    - @kgodey I hope this is fine. If you feel strongly about introducing a separate header, we could discuss on it.
 - Views are given a distinct color, even though the icon remains similar to tables.
    - I prefer the icons to look similar since views and tables share considerable similarity. I did change the color to make it more distinct, while removing the tags.
    - @zackkrida Let me know how you feel about this. 
 - The header "Tables & Views" only appears when the schema has views. If the schema does not have views, only "Tables" is displayed.
- When Views are present:
     - <img width="773" height="584" alt="Screenshot 2025-11-26 at 7 26 38 PM" src="https://github.com/user-attachments/assets/98e50968-5a82-4048-8144-98331f2239c3" />
- When Views are not present:
    - <img width="773" height="380" alt="Screenshot 2025-11-26 at 7 26 46 PM" src="https://github.com/user-attachments/assets/114b6263-9a1c-4e39-bdbd-b49b3951ee2a" />

pavish commented:
Response to @seancolsen:

> I noticed the following functionality that doesn't work but is present (and usable) in the UI:
> Rename view
> Delete view

I'd like both of these to work by fixing this on the backend. I think the error is fine for now since the publicly hosted dataset is going to non-editable and these options would appear disabled. We can handle these cases in separate smaller PRs.

> Delete column
> Add column
> Alter column data type
> Extract columns
> Set column default value

I agree these need to be hidden. I've not prioritized this since the error looked good enough and didn't consider this essential for this release. 

These scenarios were intentionally not handled for the sake of time, but based on the release timeline and QA, we can consider handling them.

zackkrida commented:
looks good to me @pavish ! we'll see what @kgodey says.

mathemancer commented:
> Response to @seancolsen:
> 
> > I noticed the following functionality that doesn't work but is present (and usable) in the UI:
> > Rename view
> > Delete view
> 
> I'd like both of these to work by fixing this on the backend. I think the error is fine for now since the publicly hosted dataset is going to non-editable and these options would appear disabled. We can handle these cases in separate smaller PRs.

Keep in mind, the bar here is not for folks to be able to actually manage views from Mathesar, it's to allow users who already have views in their database to be able to _view_ them.

## Rename view

This UI interaction is misnamed, and the part that's actually breaking is _not_ the "rename" part. That works fine, as you can see by renaming the view in the inspector from the view's page. What's breaking is changing the _description_, which happens in that same widget. The front end is sending a `patch` request that explicitly sets the description whether or not it's been changed by the user. This behavior is questionable anyway (e.g., if the user doesn't change _or even look at_ the description, should that overwrite another user's desired description change?), and I'd suggest either removing the description changing from that widget, or only sending a change of description portion of the request if the user actually edits it.

Request when changing only the name (not interacting with the description at all) from the tables and views list:

```json
[
  {
    "jsonrpc": "2.0",
    "id": 0,
    "method": "tables.patch",
    "params": {
      "database_id": 2,
      "table_oid": 19642,
      "table_data_dict": {
        "name": "Number of Movies Per Genre",
        "description": ""
      }
    }
  }
]
```

Relevant error snippet:

```
mathesar_service_dev  | psycopg.errors.WrongObjectType: "Number of Movies Per Genre" is not a table
mathesar_service_dev  | CONTEXT:  SQL statement "COMMENT ON TABLE movie_rentals."Number of Movies Per Genre" IS ''"
mathesar_service_dev  | PL/pgSQL function msar.comment_on_table(oid,text) line 9 at EXECUTE
mathesar_service_dev  | SQL statement "SELECT CASE WHEN tab_alters ? 'description'
mathesar_service_dev  |   THEN msar.comment_on_table(tab_id, tab_alters->>'description') END"
mathesar_service_dev  | PL/pgSQL function msar.alter_table(oid,jsonb) line 22 at PERFORM
```

Request when changing the name using the inspector in the view page:
```json
[
  {
    "jsonrpc": "2.0",
    "id": 0,
    "method": "tables.patch",
    "params": {
      "database_id": 2,
      "table_oid": 19642,
      "table_data_dict": {
        "name": "Number of Movies Per Genres"
      }
    }
  }
]
```

This works just fine.

For comic relief: When editing the description and the view name in the inspector without saving in between,

<img width="706" height="918" alt="image" src="https://github.com/user-attachments/assets/07647a50-0e67-4f9d-984d-7fe9c2f5dec8" />

clicking save on the view name _works_ because the front end sends the same JSON:

```json
[
  {
    "jsonrpc": "2.0",
    "id": 0,
    "method": "tables.patch",
    "params": {
      "database_id": 2,
      "table_oid": 19642,
      "table_data_dict": {
        "name": "Number of Movies Per Genres"
      }
    }
  }
]
```

Enabling editing the description of a view is (I'd argue) beyond the scope of what we want to accomplish for this release, and honestly isn't a very high priority compared to other view alterations.

## Delete view

This is also beyond the scope of what we want to accomplish for this release. When we get into manipulating views in that way, I think we should separate the namespaces on the backend (i.e., have a separate `views` namespace), and do it "properly". There are a number of reasons for that, but I'd prefer we defer that whole discussion and just agree that manipulating views is beyond our current scope.

zackkrida committed:
Merge branch 'develop' into views_fe

zackkrida commented:
@kgodey as discussed this is ready for final review. Thanks!

pavish committed:
Merge branch 'develop' into views_fe

pavish committed:
Merge branch 'develop' into views_fe

pavish committed:
Merge branch 'develop' into views_fe

pavish committed:
Hide/disable actions not supported for views

pavish committed:
Use the same icon as the table icon for views

pavish committed:
Set table inspector tab to table/view when table changes

pavish committed:
Hide join options for views

kgodey reviewed (approved):
Approved on call with @pavish.

pavish commented:
There have been some more UX changes from my call with @kgodey:

- Views use the same icon as tables, but with a different color.
- Views are displayed at the top of the list.
- Actions that currently do not work for views are hidden. (Eg., editing description, record options in inspector etc.,)
- <img width="1441" height="798" alt="Screenshot 2025-12-08 at 9 37 41 PM" src="https://github.com/user-attachments/assets/9f67a50d-5245-4e0f-a28e-8b4636a756d8" />
- <img width="1441" height="798" alt="Screenshot 2025-12-08 at 9 37 33 PM" src="https://github.com/user-attachments/assets/fcad6f35-6eda-40ae-b016-b7c6fbafab6b" />

@zackkrida @seancolsen I'm merging this PR in.

@seancolsen I made some minimal code changes for this UX, please feel free to review post-merge.
